### PR TITLE
Add cooperative GPU-sharing infrastructure via CUDA process-checkpoint

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #181215

checkpoint/cuda_checkpoint.py wraps cuCheckpointProcess{Lock,Checkpoint,
Restore,Unlock} via ctypes so a process can release its entire CUDA
context (VRAM returned to the driver) and reclaim it later with all
device pointers intact.

checkpoint/coordinator.py + checkpoint/coord_client.py + checkpoint/protocol.py
build on those primitives. Multiple torch processes register with a
single coordinator by integer rank. Exactly one rank holds the GPU at a
time; all data exchange goes through one primitive:

    prepare(send: dict[tuple[int], Tensor | None], recv: tuple[int])
        -> dict[int, Tensor | None] | None

The caller declares what tensors it is depositing for which peers, and
which peers it expects tensors from. If all recv sources have already
deposited (fast path), prepare returns the data and the caller keeps
the baton. If any source is not yet ready, prepare returns None and the
caller must call release_gpu(), which checkpoints the process, waits
for the coordinator to schedule it, and restores + returns the data.

Every PyTorch collective (send/recv, broadcast, reduce, gather, scatter,
all_gather, all_reduce, reduce_scatter, all_to_all, barrier) maps to a
single prepare call; the full table is in CoordClient.prepare's
docstring. Per-pair FIFO ordering matches PyTorch's "user ensures
collective call order" contract. Mismatches between a rank's send/recv
and its peers' are detected at prepare time and raised as
CollectiveMismatch.

checkpoint_demos/ contains three flows:
 - demo_ping/demo_pong/run_ping_pong: pre-coordinator ping-pong using
   the file-token Baton (kept for reference).
 - demo_coord_ping/demo_coord_pong/run_coord_demo: ranks 0 and 1 doing a
   2-way all-reduce via prepare(). Asserts the reduction is correct on
   both sides.
 - bench_baton/run_bench/summarize_bench: times checkpoint/restore in
   three sweeps.

Review order: cuda_checkpoint.py (driver wrappers), then protocol.py (op
names, error codes, wire format), then coordinator.py (server scheduling
+ mismatch detection), then coord_client.py (sync API + tensor codec),
then the demos/bench.

Findings from the first full perf run (H100, 5 iters per config, iter 0
dropped as warmup):

Sweep A (size, 1 tensor, symmetric):
  64 MB    ckpt  496 ms   restore  479 ms
  256 MB   ckpt  600 ms   restore  468 ms
  1 GB     ckpt  1.1 s    restore  1.2 s
  4 GB     ckpt  3.0 s    restore  1.4 s
  16 GB    ckpt  9.3 s    restore  3.9 s
  40 GB    ckpt  27.9 s   restore  11.0 s
Fixed floor around 500 ms, then roughly linear above 1 GB. Sustained ~1.5 GB/s
on checkpoint and ~4 GB/s on restore on a PCIe Gen4 link that can do ~20 GB/s,
so the driver is spending a lot of time per page, not waiting on the bus.
Restore is 2-3x faster than checkpoint at equal size.

Sweep B (count, 1 GB fixed, symmetric): essentially flat at 1.1-1.4 s ckpt /
~0.85 s restore for every count from 1 to 100000 tensors. The caching
allocator coalesces small allocations into larger segments and checkpoint
operates on segments, so the user-visible tensor count is invisible to
checkpoint cost.

Sweep C (symmetry, ping fixed at 1 GB / 100 tensors):
  pong 64 MB       ping ckpt 1.10 s   restore 827 ms
  pong 1 GB        ping ckpt 1.29 s   restore 891 ms
  pong 16 GB       ping ckpt 1.91 s   restore 859 ms
  pong 1 GB / 10k  ping ckpt 1.48 s   restore 931 ms
Cross-process coupling is real: ping's own 1 GB checkpoint takes ~1.7x longer
when pong is holding 16 GB, even though only ping is being checkpointed at
that moment. Likely contention on shared driver staging or a global lock.
Restore times are not materially affected.

Authored with Claude.

### CODE REVIEW

File: test/distributed/_experimental/test_torchmux.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+from torch.distributed._experimental.torchmux.coordinator import (
+    _PeerGone,
+    _Prepare,
+    Coordinator,
+)
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+def _start_thread(fn):
+    q = queue.Queue()
```
Comment (bobrenjc93) [inline test/distributed/_experimental/test_torchmux.py:31]: for my own understanding, what is the point of the queue? why not just return the tuple?

File: test/distributed/_experimental/test_torchmux.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+        try:
+            t0, q0 = _start_thread(lambda: c0.register(0))
+            t1, q1 = _start_thread(lambda: c1.register(1))
+
+            self._thread_result(t0, q0)
+
+            # rank 1 is still blocked in register
+            self.assertTrue(t1.is_alive())
+
+            # rank 0 does a collective that requires rank 1's data — it blocks,
```
Comment (bobrenjc93) [inline test/distributed/_experimental/test_torchmux.py:108]: these are nice tests

File: torch/distributed/_experimental/torchmux/backend.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        t = tensors[0]
+        others = self._others
+        received = self._exchange(
+            send={others: t},
+            recv=others,
+            tensors=(t,),
+        )
+        op = opts.reduceOp
+        for peer_t in received.values():
+            assert peer_t is not None
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/backend.py:204]: remember to fix lint

File: torch/distributed/_experimental/torchmux/backend.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+        tensors, opts = args
+        t = tensors[0]
+        others = self._others
+        received = self._exchange(
+            send={others: t},
+            recv=others,
+            tensors=(t,),
+        )
+        op = opts.reduceOp
+        for peer_t in received.values():
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/backend.py:203]: I wonder if we should include rank as part of received so when users write their own custom reduce meta kernels, they can use rank to do topology estimations

File: torch/distributed/_experimental/torchmux/coord_client.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+
+class CollectiveMismatch(CoordClientError):
+    """Raised when this rank's prepare is inconsistent with a peer's."""
+
+
+# ---- Tensor serialization (raw-bytes fast path) ----
+
+
+def _serialize_tensor(tensor: object) -> tuple[TensorHeader, bytes]:
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/coord_client.py:59]: i'm surprised we don't have an existing utility for this

File: torch/distributed/_experimental/torchmux/coord_client.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+            finally:
+                self._loop.close()
+
+        self._thread = threading.Thread(
+            target=run_loop, name="CoordClient", daemon=True
+        )
+        self._thread.start()
+        self._ready.wait()
+
+    def _run_coro(self, coro: Any, *, timeout: float | None = None) -> Any:
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/coord_client.py:190]: it's not immediately obvious what coro stands for, maybe expand to full name?

Comment (bobrenjc93) [reply torch/distributed/_experimental/torchmux/coord_client.py:190]: oh coroutine...

File: torch/distributed/_experimental/torchmux/examples/allreduce.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ap = argparse.ArgumentParser()
+ap.add_argument(
+    "--backend",
+    default="nccl",
+    help="Distributed backend: nccl, gloo, torchmux, etc.",
+)
+args = ap.parse_args()
+
+if args.backend == "torchmux":
+    import torch.distributed._experimental.torchmux
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/examples/allreduce.py:27]: is the plan to have torchrun hot swap out args.backend with torchmux? what if user code hardcodes the backend? do we still swap out under the hood?

File: torch/distributed/_experimental/torchmux/examples/allreduce.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+dist.init_process_group(backend=args.backend)
+rank = dist.get_rank()
+world = dist.get_world_size()
+
+if args.backend == "nccl":
+    device = f"cuda:{rank}"
+else:
+    device = "cpu"
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/examples/allreduce.py:36]: i'm a little confused here, is the expecetd use case for args.backend == "torchmux"? if so wouldn't we want torchmux backend to have tensors also go on cuda?

File: torch/distributed/_experimental/torchmux/backend.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+def _run_coordinator_in_thread() -> str:
+    """Start the coordinator server in a daemon thread. Returns the TCP address."""
+    from .coordinator import Coordinator
+
+    ready = threading.Event()
+    addr_box: list[str] = []
+
+    async def _serve() -> None:
+        coord = Coordinator()
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/backend.py:95]: interesting, i guess this approach should also scale to multiple nodes right? presumably you'd need to change the default from 127.0.0.1 to 0.0.0.0

File: torch/distributed/_experimental/torchmux/coord_client.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    def register(self, rank: int) -> None:
+        """Register with the coordinator.
+
+        Rank 0 returns immediately once registered.  Later ranks block until
+        their rank-ordered turn and the active rank yields.
+        """
+        if not isinstance(rank, int):
+            raise TypeError(f"rank must be int, got {type(rank).__name__}")
+        msg: JsonDict = {"op": OP_REGISTER, "rank": rank}
+        self._run_coro(self._engine.rpc(msg))
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/coord_client.py:217]: how does error handling work here? if there is a network partition, etc. since this is a coroutine... what happens?

File: torch/distributed/_experimental/torchmux/backend.py [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+    def getBackendName(self) -> str:
+        return "torchmux"
+
+    def _exchange(
+        self,
+        send: dict[tuple[int, ...], torch.Tensor | None],
+        recv: tuple[int, ...],
+        tensors: object = (),
+        force_cuda: bool = False,
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/backend.py:153]: it's unclear why this flag is named force_cuda instead of force_release_gpu. maybe add a comment?

File: torch/distributed/_experimental/torchmux/coord_client.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+        return self._decode_recv(_recv_entries(resp_hdr), resp_payload)
+
+    def _wait_for_recv_raw(self) -> tuple[JsonDict, bytes]:
+        """Send release_gpu and return the raw framed response."""
+        resp_hdr, resp_payload = self._run_coro(
+            self._engine.rpc({"op": OP_RELEASE_GPU})
+        )
+        return resp_hdr, resp_payload
+
+    def done(self) -> None:
```
Comment (bobrenjc93) [inline torch/distributed/_experimental/torchmux/coord_client.py:329]: can you add a docblock here? it's not immediately obvious why this is necessary? eg. why isn't prepare alone sufficient
